### PR TITLE
Add support for grant on database propagation from non-main databases

### PR DESCRIPTION
**Pull request by gurkanindibay**
DESCRIPTION: Adds support for distributed `GRANT .. ON DATABASE TO USER` commands from the databases where Citus is not installed

### CODE REVIEW

File: src/test/regress/expected/grant_on_database_propagation_from_non_maindb.out
Comment (onurctirtir) [inline src/test/regress/expected/grant_on_database_propagation_from_non_maindb.out:383]: Comment says that this should fail. Is it an outdated comment?

File: src/test/regress/expected/grant_on_database_propagation_from_non_maindb.out
Comment (onurctirtir) [inline src/test/regress/expected/grant_on_database_propagation_from_non_maindb.out:1]: Tests really look great!

And to prevent this test output getting flaky in future, can we add some order by's to `select check_database_privileges..` stmts that return 12 rows? The ones that return 3 rows are already fine.

File: src/test/regress/sql/grant_on_database_propagation_from_non_maindb.sql
Comment (onurctirtir) [inline src/test/regress/sql/grant_on_database_propagation_from_non_maindb.sql:10]: let's delete trailing newlines in the file

File: src/test/regress/sql/metadata_sync_from_non_maindb.sql
Comment (onurctirtir) [inline src/test/regress/sql/metadata_sync_from_non_maindb.sql:1]: let's delete unnecessary newline additions in the file

File: src/backend/distributed/commands/utility_hook.c
Comment (onurctirtir) [inline src/backend/distributed/commands/utility_hook.c:133]: where do we make use of third field of those arrays in the code?

File: src/test/regress/sql/grant_on_database_propagation_from_non_maindb.sql
Comment (gurkanindibay) [reply src/test/regress/sql/grant_on_database_propagation_from_non_maindb.sql:9]: Removed

File: src/test/regress/sql/metadata_sync_from_non_maindb.sql
Comment (gurkanindibay) [reply src/test/regress/sql/metadata_sync_from_non_maindb.sql:1]: Removed

File: src/backend/distributed/commands/utility_hook.c
Comment (gurkanindibay) [reply src/backend/distributed/commands/utility_hook.c:133]: Disappeared after the merge. I re-added the logic there

File: src/test/regress/expected/grant_on_database_propagation_from_non_maindb.out
Comment (gurkanindibay) [reply src/test/regress/expected/grant_on_database_propagation_from_non_maindb.out:1]: check_database_privileges function gets array input and process the array with the given order. So the order depends on the given array input.
Since we do not rely on any external data other than inputs and we process the inputs with the given order, there is no need for ordering

File: src/test/regress/expected/grant_on_database_propagation_from_non_maindb.out
Comment (gurkanindibay) [reply src/test/regress/expected/grant_on_database_propagation_from_non_maindb.out:383]: You're right outdated comment

File: src/test/regress/expected/grant_on_database_propagation_from_non_maindb.out
Comment (onurctirtir) [reply src/test/regress/expected/grant_on_database_propagation_from_non_maindb.out:1]: makes sense then

File: src/backend/distributed/commands/utility_hook.c
Comment (onurctirtir) [inline src/backend/distributed/commands/utility_hook.c:1727]: To me, this block feels a bit hard to maintain. Would you like to cherry-pick the changes from https://github.com/citusdata/citus/commit/6d73c8381cf44d248229a5d5f9e1ea1632f82058 instead?

This one makes use of some function pointers to avoid clutter in IsStatementSupportedFromNonMainDb.

Comment (onurctirtir) [reply src/backend/distributed/commands/utility_hook.c:1727]: Apart from that, those non-main db related code definitely need to be moved into a new file soon but I'm refraining from doing this now to avoid further merge conflicts.

However, we should do that maybe after merging currently active PRs touching this code-path.

File: src/test/regress/sql/metadata_sync_from_non_maindb.sql
Comment (onurctirtir) [reply src/test/regress/sql/metadata_sync_from_non_maindb.sql:1]: there are still some in this file

**Review by onurctirtir — COMMENTED** [review]
LGTM after addressing those last couple of comments

File: src/backend/distributed/commands/utility_hook.c
Comment (gurkanindibay) [reply src/backend/distributed/commands/utility_hook.c:1727]: Totally agree with you, after maturization of this structure, we can easily move it to another file/files

Comment (gurkanindibay) [reply src/backend/distributed/commands/utility_hook.c:1727]: I like the function pointers solution. I'll cherry-pick and make required changes if it is required
Thanks

Comment (gurkanindibay) [reply src/backend/distributed/commands/utility_hook.c]: Cherry-picked. Thanks for the contribution

**Review by onurctirtir — APPROVED** [review]
Please consider the last comment before merging the PR and thanks for improving the infra & adding support for another class of commands from non-main dbs!

**Comment by gurkanindibay** [comment]
> Please consider the last comment before merging the PR and thanks for improving the infra & adding support for another class of commands from non-main dbs!

Interesting. Probably I worked on non-freshed page on Github so that I didn't notice your last commit. Thanks for the heads up

**Pull request merged**
